I think that particular failure is benign and will resolve itself if we update the PR. However there's another problem, namely the tests now fail on TPU. The two failures are:

```
LaxControlFlowTest.test_custom_root_scalar_binary_search

Traceback (most recent call last):
...
  File "jax/tests/lax_control_flow_test.py", line 2079, in test_custom_root_scalar
    self.assertAllClose(results, inputs ** 1.5, check_dtypes=False, atol=1e-6, rtol=1e-6)
...

AssertionError: 
Not equal to tolerance rtol=1e-06, atol=1e-06

Mismatched elements: 1 / 2 (50%)
Max absolute difference: 0.00037384
Max relative difference: 3.343618e-05
 x: array([ 8.     , 11.18034], dtype=float32)
 y: array([ 8.      , 11.180714], dtype=float32)
```

and
```
LaxControlFlowTest.test_custom_root_scalar_newton_raphson

Traceback (most recent call last):
 ...
  File "jax/tests/lax_control_flow_test.py", line 2079, in test_custom_root_scalar
    self.assertAllClose(results, inputs ** 1.5, check_dtypes=False, atol=1e-6, rtol=1e-6)
...
AssertionError: 
Not equal to tolerance rtol=1e-06, atol=1e-06

Mismatched elements: 1 / 2 (50%)
Max absolute difference: 0.00037384
Max relative difference: 3.343618e-05
 x: array([ 8.     , 11.18034], dtype=float32)
 y: array([ 8.      , 11.180714], dtype=float32)
```

I suspect the easiest thing to do is just to relax the test tolerances a little bit?

(Making an update to the PR should also retrigger the CI build that failed.)

--- a/docs/notebooks/Custom_derivative_rules_for_Python_code.ipynb
+++ b/docs/notebooks/Custom_derivative_rules_for_Python_code.ipynb
@@ -1070,7 +1070,7 @@
     "id": "HowvqayEuy-H"
    },
    "source": [
-    "A limitation to this approach is that the argument `f` can't close over any values involved in differentiation. That is, you might notice that we kept the parameter `a` explicit in the argument list of `fixed_point`. While other JAX mechanisms can handle closed-over transformation-traced values in the arguments to higher-order functions (as is done for the control flow primitives like `lax.cond`, `lax.scan`, and `lax.while_loop` itself), `jax.custom_vjp` used as above cannot. A `fixed_point` routine that used a bit more of JAX's internals could have a more convenient and robust API."
+    "A limitation to this approach is that the argument `f` can't close over any values involved in differentiation. That is, you might notice that we kept the parameter `a` explicit in the argument list of `fixed_point`. For this use case, consider using the low-level primitive `lax.custom_root`, which allows for deriviatives in closed-over variables with custom root-finding functions."
    ]
   },
   {

--- a/docs/notebooks/Custom_derivative_rules_for_Python_code.md
+++ b/docs/notebooks/Custom_derivative_rules_for_Python_code.md
@@ -560,7 +560,7 @@ print(grad(grad(jnp.sqrt))(2.))
 
 +++ {"id": "HowvqayEuy-H"}
 
-A limitation to this approach is that the argument `f` can't close over any values involved in differentiation. That is, you might notice that we kept the parameter `a` explicit in the argument list of `fixed_point`. While other JAX mechanisms can handle closed-over transformation-traced values in the arguments to higher-order functions (as is done for the control flow primitives like `lax.cond`, `lax.scan`, and `lax.while_loop` itself), `jax.custom_vjp` used as above cannot. A `fixed_point` routine that used a bit more of JAX's internals could have a more convenient and robust API.
+A limitation to this approach is that the argument `f` can't close over any values involved in differentiation. That is, you might notice that we kept the parameter `a` explicit in the argument list of `fixed_point`. For this use case, consider using the low-level primitive `lax.custom_root`, which allows for deriviatives in closed-over variables with custom root-finding functions.
 
 +++ {"id": "Dr0aNkBslfQf"}
 

--- a/jax/_src/lax/control_flow.py
+++ b/jax/_src/lax/control_flow.py
@@ -2357,18 +2357,6 @@ def _promote_weak_typed_inputs(in_vals, in_avals, out_avals):
     in_vals[i] = lax.convert_element_type(in_vals[i], new_dtype)
   return in_vals, True
 
-def _stop_gradient_fun(f):
-  """Create a version of f() that stops all gradients."""
-  def wrapper(*args, **kwargs):
-    args_flat, in_args_tree = tree_flatten((args, kwargs))
-    args_avals = tuple(_map(_abstractify, args_flat))
-    g = lambda a, b: f(*a, **b)
-    jaxpr, consts, out_tree = _initial_style_jaxpr(g, in_args_tree, args_avals)
-    all_args = _map(lax.stop_gradient, (*consts, *args_flat))
-    out = core.jaxpr_as_fun(jaxpr)(*all_args)
-    return tree_unflatten(out_tree, out)
-  return wrapper
-
 
 _RootTuple = collections.namedtuple('_RootTuple', 'f, solve, l_and_s')
 
@@ -2426,7 +2414,7 @@ def custom_root(f, initial_guess, solve, tangent_solve, has_aux=False):
   _check_tree("f", "initial_guess", out_tree, in_tree, False)
 
   solve_jaxpr, solve_consts, solution_tree = _initial_style_jaxpr(
-      partial(solve, _stop_gradient_fun(f)), in_args_tree, guess_avals)
+      partial(solve, f), in_args_tree, guess_avals)
   _check_tree("solve", "initial_guess", solution_tree, in_tree, has_aux)
 
   def linearize_and_solve(x, b):

--- a/tests/lax_control_flow_test.py
+++ b/tests/lax_control_flow_test.py
@@ -95,6 +95,52 @@ ignore_jit_of_pmap_warning = partial(
   jtu.ignore_warning, message=".*jit-of-pmap.*")
 
 
+# Simple optimization routine for testing custom_root
+def binary_search(func, x0, low=0.0, high=100.0):
+  del x0  # unused
+
+  def cond(state):
+    low, high = state
+    midpoint = 0.5 * (low + high)
+    return (low < midpoint) & (midpoint < high)
+
+  def body(state):
+    low, high = state
+    midpoint = 0.5 * (low + high)
+    update_upper = func(midpoint) > 0
+    low = jnp.where(update_upper, low, midpoint)
+    high = jnp.where(update_upper, midpoint, high)
+    return (low, high)
+
+  solution, _ = lax.while_loop(cond, body, (low, high))
+  return solution
+
+# Optimization routine for testing custom_root.
+def newton_raphson(func, x0):
+  tol = 1e-16
+  max_it = 20
+
+  fx0, dfx0 = func(x0), jax.jacobian(func)(x0)
+  initial_state = (0, x0, fx0, dfx0)  # (iteration, x, f(x), grad(f)(x))
+
+  def cond(state):
+    it, _, fx, _ = state
+    return (jnp.max(jnp.abs(fx)) > tol) & (it < max_it)
+
+  def body(state):
+    it, x, fx, dfx = state
+    step = jnp.linalg.solve(
+      dfx.reshape((-1, fx.size)), fx.ravel()
+    ).reshape(fx.shape)
+    x_next = x - step
+    fx, dfx = func(x_next), jax.jacobian(func)(x_next)
+    return (it + 1, x_next, fx, dfx)
+
+  _, x, _, _ = lax.while_loop(cond, body, initial_state)
+
+  return x
+
+
 class LaxControlFlowTest(jtu.JaxTestCase):
 
   def setUp(self):
@@ -2007,33 +2053,19 @@ class LaxControlFlowTest(jtu.JaxTestCase):
 
     jax.grad(lambda x: jit_run_scan(x))(0.)  # doesn't crash
 
-  def test_custom_root_scalar(self):
+  @parameterized.named_parameters(
+      {"testcase_name": "binary_search", "solve_method": binary_search},
+      {"testcase_name": "newton_raphson", "solve_method": newton_raphson},
+  )
+  def test_custom_root_scalar(self, solve_method):
 
     def scalar_solve(f, y):
       return y / f(1.0)
 
-    def binary_search(func, x0, low=0.0, high=100.0):
-      del x0  # unused
-
-      def cond(state):
-        low, high = state
-        midpoint = 0.5 * (low + high)
-        return (low < midpoint) & (midpoint < high)
-
-      def body(state):
-        low, high = state
-        midpoint = 0.5 * (low + high)
-        update_upper = func(midpoint) > 0
-        low = jnp.where(update_upper, low, midpoint)
-        high = jnp.where(update_upper, midpoint, high)
-        return (low, high)
-
-      solution, _ = lax.while_loop(cond, body, (low, high))
-      return solution
-
     def sqrt_cubed(x, tangent_solve=scalar_solve):
       f = lambda y: y ** 2 - x ** 3
-      return lax.custom_root(f, 0.0, binary_search, tangent_solve)
+      # Note: Nonzero derivative at x0 required for newton_raphson
+      return lax.custom_root(f, 1.0, solve_method, tangent_solve)
 
     value, grad = jax.value_and_grad(sqrt_cubed)(5.0)
     self.assertAllClose(value, 5 ** 1.5, check_dtypes=False, rtol=1e-6)
@@ -2044,7 +2076,11 @@ class LaxControlFlowTest(jtu.JaxTestCase):
 
     inputs = jnp.array([4.0, 5.0])
     results = jax.vmap(sqrt_cubed)(inputs)
-    self.assertAllClose(results, inputs ** 1.5, check_dtypes=False)
+    self.assertAllClose(
+      results, inputs ** 1.5, check_dtypes=False,
+      atol={jnp.float32: 1e-3, jnp.float64: 1e-6},
+      rtol={jnp.float32: 1e-3, jnp.float64: 1e-6},
+    )
 
     results = jax.jit(sqrt_cubed)(5.0)
     self.assertAllClose(
@@ -2072,6 +2108,30 @@ class LaxControlFlowTest(jtu.JaxTestCase):
     actual = jax.jit(linear_solve)(a, b)
     expected = jnp.linalg.solve(a, b)
     self.assertAllClose(expected, actual)
+
+  def test_custom_root_vector_nonlinear(self):
+
+    def nonlinear_func(x, y):
+      # func(x, y) == 0 if and only if x == y.
+      return (x - y) * (x**2 + y**2 + 1)
+
+    def tangent_solve(g, y):
+      return jnp.linalg.solve(
+        jax.jacobian(g)(y).reshape(-1, y.size),
+        y.ravel()
+      ).reshape(y.shape)
+
+    def nonlinear_solve(y):
+      f = lambda x: nonlinear_func(x, y)
+      x0 = -jnp.ones_like(y)
+      return lax.custom_root(f, x0, newton_raphson, tangent_solve)
+
+    y = self.rng().randn(3, 1)
+    jtu.check_grads(nonlinear_solve, (y,), order=2,
+                rtol={jnp.float32: 1e-2, jnp.float64: 1e-3})
+
+    actual = jax.jit(nonlinear_solve)(y)
+    self.assertAllClose(y, actual, rtol=1e-5, atol=1e-5)
 
   def test_custom_root_with_custom_linear_solve(self):
 

